code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/dao/impl/ServiceExposeMapDaoImpl.java

```diff
@@ -42,4 +67,74 @@ public Instance getServiceInstance(long serviceId, String instanceName) {
+
+    @Override
+    public Instance createServiceInstance(Map<Object, Object> properties, Service service, String instanceName) {
+        Map<String, Object> props = objectManager.convertToPropertiesFor(Instance.class,
+                properties);
+            Instance instance = objectManager.create(Instance.class, props);
+            objectManager.create(ServiceExposeMap.class,
+                    SERVICE_EXPOSE_MAP.INSTANCE_ID,
+                    instance.getId(), SERVICE_EXPOSE_MAP.SERVICE_ID, service.getId());
+            objectProcessManager.scheduleStandardProcess(StandardProcess.CREATE, instance, null);
+        return objectManager.reload(instance);
+    }
+
+    @Override
+    public List<? extends Instance> listServiceInstances(long serviceId) {
+        return create()
+                .select(INSTANCE.fields())
+                .from(INSTANCE)
+                .join(SERVICE_EXPOSE_MAP)
+                .on(SERVICE_EXPOSE_MAP.INSTANCE_ID.eq(INSTANCE.ID)
+                        .and(SERVICE_EXPOSE_MAP.SERVICE_ID.eq(serviceId))
+                        .and(SERVICE_EXPOSE_MAP.STATE.in(CommonStatesConstants.ACTIVATING,
+                                CommonStatesConstants.ACTIVE, CommonStatesConstants.REQUESTED)))
+                .where(INSTANCE.REMOVED.isNull())
+                .fetchInto(InstanceRecord.class);
+    }
+
+    @Override
+    public List<? extends ServiceExposeMap> listRemovedInstanceServiceMaps(long serviceId) {
+        return create()
+                .select(SERVICE_EXPOSE_MAP.fields())
+                .from(SERVICE_EXPOSE_MAP)
+                .join(INSTANCE)
+                .on(SERVICE_EXPOSE_MAP.INSTANCE_ID.eq(INSTANCE.ID)
+                        .and(SERVICE_EXPOSE_MAP.SERVICE_ID.eq(serviceId))
+                        .and(SERVICE_EXPOSE_MAP.REMOVED.isNull()))
+                .where(INSTANCE.STATE.eq(CommonStatesConstants.REMOVED))
+                .fetchInto(ServiceExposeMapRecord.class);
+    }
+
+    @Override
+    public void updateServiceName(Service service) {
+        List<? extends Instance> instances = updateInstanceNames(service);
+        for (Instance instance : instances) {
+            objectManager.persist(instance);
+        }
+    }
+
+    private List<? extends Instance> updateInstanceNames(Service service) {
+        List<? extends Instance> instances = listServiceInstances(service.getId());
+        int i = 1;
+        for (Instance instance : instances) {
+            instance.setName(sdService.generateServiceInstanceName(service, i));
+            i++;
+        }
+        return instances;
+    }
+
+    @Override
+    public void updateEnvironmentName(Environment env) {
+        List<? extends Service> services = objectManager.mappedChildren(env, Service.class);
+        List<Instance> instances = new ArrayList<>();
+        for (Service service : services) {
+            if (service.getRemoved() == null && !service.getState().equalsIgnoreCase(CommonStatesConstants.REMOVED))
```

Comment:
I wonder should we filter out CommonStatesConstants.REMOVING as well.  Probably not a big deal since we're just updating the instance names.